flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -3340,6 +3341,99 @@ class ArrayExprLowering {
     return fir::ArrayBoxValue(tempRes, dest.getExtents());
   }
 
+  static ExtValue lowerLazyArrayExpression(
+      Fortran::lower::AbstractConverter &converter,
+      Fortran::lower::SymMap &symMap, Fortran::lower::StatementContext &stmtCtx,
+      const Fortran::evaluate::Expr<Fortran::evaluate::SomeType> &expr,
+      mlir::Value var) {
+    ArrayExprLowering ael{converter, stmtCtx, symMap};
+    return ael.lowerLazyArrayExpression(expr, var);
+  }
+
+  ExtValue lowerLazyArrayExpression(
+      const Fortran::evaluate::Expr<Fortran::evaluate::SomeType> &expr,
+      mlir::Value var) {
+    auto loc = getLoc();
+    // Once the loop extents have been computed, which may require being inside
+    // some explicit loops, lazily allocate the expression on the heap.
+    ccPrelude = [=](llvm::ArrayRef<mlir::Value> shape) -> mlir::Value {
+      auto load = builder.create<fir::LoadOp>(loc, var);
+      auto eleTy = fir::unwrapRefType(load.getType());
+      auto unknown = fir::SequenceType::getUnknownExtent();
+      fir::SequenceType::Shape extents(shape.size(), unknown);
+      auto seqTy = fir::SequenceType::get(extents, eleTy);
+      auto toTy = fir::HeapType::get(seqTy);
+      auto castTo = builder.createConvert(loc, toTy, load);
+      auto cmp = builder.genIsNull(loc, castTo);
+      auto ifOp = builder.create<fir::IfOp>(loc, cmp, /*withElseRegion=*/false);
+      auto insPt = builder.saveInsertionPoint();
+      builder.setInsertionPointToStart(&ifOp.thenRegion().front());
+      auto mem = builder.create<fir::AllocMemOp>(loc, seqTy, ".lazy.mask",
+                                                 llvm::None, shape);
+      auto uncast = builder.createConvert(loc, load.getType(), mem);
+      builder.create<fir::StoreOp>(loc, uncast, var);
+      builder.restoreInsertionPoint(insPt);
+      return mem;
```

Comment:
Is it OK to return mem here now that we are outside of the if region (it seems nothing is actually done with it other than an assert, but it seems a bit wrong) ?
Also, maybe you could use  builder.genIfThen(loc, cdt).genThen(lambda).end() to help delimiting the code here.

Reply:
This is sort of a leftover from earlier. It didn't hurt anything, so I left it.

Right. I forgot you added the IfOp generator.